packages/contracts/contracts/UnipigTransitionEvaluator.sol

```diff
@@ -0,0 +1,331 @@
+pragma solidity ^0.5.0;
+pragma experimental ABIEncoderV2;
+
+/* Internal Imports */
+import {DataTypes as dt} from "./DataTypes.sol";
+import {TransitionEvaluator} from "./TransitionEvaluator.sol";
+
+contract UnipigTransitionEvaluator is TransitionEvaluator {
+    bytes32 ZERO_BYTES32 = 0x0000000000000000000000000000000000000000000000000000000000000000;
+    address UNISWAP_ADDRESS = 0x0000000000000000000000000000000000000000;
+    uint UNISWAP_SLOT_INDEX = 0;
+    uint UNISWAP_FEE_IN_BIPS = 30;
+    // Transition Types
+    uint CREATE_AND_TRANSFER_TYPE = 0;
+    uint TRANSFER_TYPE = 1;
+    uint SWAP_TYPE = 2;
+
+    function evaluateTransition(
+        bytes calldata _transition,
+        dt.StorageSlot[2] calldata _storageSlots
+    ) external view returns(bytes32[2] memory) {
+        // Convert our inputs to memory
+        bytes memory transition = _transition;
+        dt.StorageSlot[2] memory storageSlots = _storageSlots;
+        // Determine the transition type
+        uint transitionType = inferTransitionType(transition);
+        // And initalize updatedStorage which will contain the new storage values
+        dt.Storage[2] memory updatedStorage;
+        // Apply the transition and record the resulting storage slots
+        if (transitionType == CREATE_AND_TRANSFER_TYPE) {
+            dt.CreateAndTransferTransition memory createAndTransfer = decodeCreateAndTransferTransition(transition);
+            updatedStorage = applyCreateAndTransferTransition(createAndTransfer, storageSlots);
+        } else if (transitionType == TRANSFER_TYPE) {
+            dt.TransferTransition memory transfer = decodeTransferTransition(transition);
+            updatedStorage = applyTransferTransition(transfer, storageSlots);
+        } else if (transitionType == SWAP_TYPE) {
+            dt.SwapTransition memory swap = decodeSwapTransition(transition);
+            updatedStorage = applySwapTransition(swap, storageSlots);
+        } else {
+            revert("Transition type not recognized!");
+        }
+        // Return the hash of both storage (leaf nodes to insert into the tree)
+        bytes32[2] memory outputs;
+        outputs[0] = getStorageHash(updatedStorage[0]);
+        outputs[1] = getStorageHash(updatedStorage[1]);
+        return outputs;
+    }
+
+    function verifyEcdsaSignature(bytes memory _signature, bytes32 _hash, address _pubkey) private pure returns(bool) {
+        return true;
+    }
+
+    /**
+     * Return the tx type inferred by the length of bytes
+     */
+    function inferTransitionType(
```

Comment:
Goes without saying, but we'll need to be darn sure of these lengths or run the risk of this contract locking funds.